faunadb/stream_events.go

```diff
@@ -0,0 +1,165 @@
+package faunadb
+
+import (
+	"fmt"
+)
+
+// StreamEventType is a stream eveny type
+type StreamEventType = string
+
+const (
+	// ErrorEventT is the stream error event type
+	ErrorEventT StreamEventType = "error"
+
+	// HistoryRewriteEventT is the stream history rewrite event type
+	HistoryRewriteEventT StreamEventType = "history_rewrite"
+
+	// StartEventT is the stream start event type
+	StartEventT StreamEventType = "start"
+
+	// VersionEventT is the stream version event type
+	VersionEventT StreamEventType = "version"
+)
+
+// StreamEvent represents a stream event with a `type` and `txn`
+type StreamEvent interface {
+	Type() StreamEventType
+	Txn() int64
+	String() string
+}
+
+// StartEvent emitted when a valid stream subscription begins.
+// Upcoming events are guaranteed to have transaction timestamps equal to or greater than
+// the stream's start timestamp.
+type StartEvent struct {
+	StreamEvent
+	txn   int64
+	event int64
+}
+
+func unMarshalStreamEvent(data Obj) (evt StreamEvent, err error) {
+	switch StreamEventType(data["type"].(StringV)) {
+	case StartEventT:
+		evt = StartEvent{
+			txn:   int64(data["txn"].(LongV)),
+			event: int64(data["event"].(LongV)),
+		}
+	case VersionEventT:
+		evt = VersionEvent{
+			txn:   int64(data["txn"].(LongV)),
+			event: data["event"].(ObjectV),
+		}
+	case ErrorEventT:
+		evt = ErrorEvent{
+			txn: int64(data["txn"].(LongV)),
+			err: errorFromStreamError(data["event"].(ObjectV)),
+		}
+
+	case HistoryRewriteEventT:
+		evt = HistoryRewriteEvent{
+			txn:   int64(data["txn"].(LongV)),
+			event: data["event"].(ObjectV),
+		}
+	}
+	return
+}
+
+// Type returns the stream event type
+func (event StartEvent) Type() StreamEventType {
+	return StartEventT
+}
+
+// Txn returns the stream event timestamp
+func (event StartEvent) Txn() int64 {
+	return event.txn
+}
+
+// Event returns the stream event as a `f.ObjectV`
+func (event StartEvent) Event() int64 {
+	return event.event
+}
+
+func (event StartEvent) String() string {
+	return fmt.Sprintf("StartEvent{event=%d, txn=%d} ", event.Event(), event.Txn())
+}
+
+// VersionEvent represents a version event that occurs upon any
+// modifications to the current state of the subscribed document.
+type VersionEvent struct {
+	StreamEvent
+	txn   int64
+	event ObjectV
```

Comment:
User's aren't suppose to handle concrete values. The `Value` decoding API is suppose to help them decode fauna values to domain objects seemly.

```suggestion
	event Value
```

Reply:
Done